pkg/GAPJulia/JuliaInterface/julia/gap.jl

```diff
@@ -68,3 +68,99 @@ for (left, right) in typecombinations
+Can be used for creating GAP literals directly from Julia.
+
+    julia> @gap (1,2,3)
+    GAP: (1,2,3)
+    julia> @gap SymmetricGroup(3)
+    GAP: SymmetricGroup( [ 1 .. 3 ] )
+    julia> @gap(SymmetricGroup)(3)
+    GAP: SymmetricGroup( [ 1 .. 3 ] )
+
+Note that the last two examples have a slight syntactical, and therefore also
+a semantical difference. The first one executes the string `SymmetricGroup(3)`
+directly inside GAP. The second example returns the function `SymmetricGroup`
+via `@gap(SymmetricGroup)`, then calls that function with the argument `3`.
+"""
+macro gap(str)
+    return EvalString(string(str))
+end
+
+export @gap
+
+"""
+    LoadPackageAndExposeGlobals(package::String, mod::String; all_globals::Bool = false)
+    LoadPackageAndExposeGlobals(package::String, mod::Module = Main; all_globals::Bool = false)
+
+`LoadPackageAndExposeGlobals` loads `package` into GAP via `LoadPackage`,
+and stores all newly defined GAP globals as globals in the module `mod`. If `mod` is
+a string, the function creates a new module, if `mod` is a Module, it uses `mod` directly.
+
+The function is intended to be used for creating mock modules for GAP packages.
+If you load the package `CAP` via
+
+    LoadPackageAndExposeGlobals( "CAP", "CAP" )
+
+you can use CAP commands via
+
+    CAP.PreCompose( a, b )
+
+"""
+function LoadPackageAndExposeGlobals(package::String, mod::String; all_globals::Bool = false)
+    mod_sym = Symbol(mod)
+    Base.MainInclude.eval(:(
+        module $(mod_sym)
+            import GAP
+        end
+    ))
+    ## Adds the new module to the Main module, so it is directly accessible in the julia REPL
+    mod_mod = Base.MainInclude.eval(:(Main.$(mod_sym)))
+
+    ## We need to call `invokelatest` as the module `mod_mod` was only created during the
+    ## call of this function in a different module, so its world age is higher than the
+    ## function calls world age.
+    Base.invokelatest(LoadPackageAndExposeGlobals, package, mod_mod; all_globals = all_globals)
+end
+
+function LoadPackageAndExposeGlobals(package::String, mod::Module; all_globals::Bool = false)
+    current_gvar_list = EvalString("ShallowCopy(NamesGVars())")
+    load_package = EvalString("LoadPackage(\"$package\")")
+    if load_package == Globals.fail
+        error("cannot load package $package")
+    end
+    new_gvar_list = nothing
+    if all_globals
+        new_gvar_list = Globals.NamesGVars()
+    else
+        new_gvar_list = Globals.Difference(Globals.NamesGVars(),current_gvar_list)
+    end
+    new_symbols = gap_to_julia(Array{Symbol,1},new_gvar_list)
+    for sym in new_symbols
+        try
+            mod.eval(:(
+                $(sym)=GAP.Globals.$(sym)
+            ))
+        catch
+        end
+    end
+end
+
+export LoadPackageAndExposeGlobals
+
+function Display(x::MPtr)
+    local_var = "julia_gap_display_tmp"
+    AssignGlobalVariable(local_var,x)
+    xx = EvalStringEx("Display($local_var);")[1]
```

Comment:
Huh, it shouldn't be necessary to jump through such hoops. With that, I don't mean there is a way to avoid it right now (sadly, I fear there is none, so your hack is needed for now). But rather, that there *should* be a way. At least for `Print`, you can use `PrintTo` to redirect to a file or stream. So perhaps we need to add `ViewTo` and DisplayTo` functions to GAP. And/or provide a more general mechanism to redirect the output.

Reply:
The GAP package GAPDoc provides functions `StringPrint` and `StringView` (based on `PrintTo1`) which create strings for a given object that describe the `Print` and `View` output for this object, respectively.
An analogous `StringDisplay` could be provided, and these functions could be taken for producing the strings needed by Julia and Jupyter notebooks and the like, at least temporarily.

It is not obvious for me which strings one really wants for a given GAP object.
The above functions create strings with line breaks according to the current line length.
In the case of `Display`, I would say that this is fine:
Those objects for which `Display` is most interesting require a two-dimensional formatting,
and the line length is needed to define the context.

In the case of `View` and `Print`, one can alternatively produce one string with line breaking hints `\<` and `\>` for a given object, and postpone the actual formatting, in the sense of responsive design.
(After the recent change #3440 in gap-system/gap, I am more optimistic that `ViewString` can work as a long-term solution.)

Reply:
I would honestly prefer to use something like `DisplayString`, which is already defined in GAP, but does not really provide me with the strings I really need.

This function is for a very special case only, i.e., using it from the Jupyter Notebook with Julia kernel. Of course, we can think about a different solution, and implement all kinds of backend for it in GAP, or even cleanup the whole output mess in GAP. But imho this function has a very huge advantage over all the stuff you propose: It works right now.

So, while all your comments are totally valid, and this function is really not nicely designed or anything, I would suggest we merge it now, and then later go through the whole shebang of doing what you propose. I would happily write a `DisplayTo` function for GAP, but keep in mind that this might be a non-trivial task.

Reply:
@fingolfin I have to apologize a bit for my whitty comment. I did not read that you also probably not think there is a better ad-hoc solution in GAP. So okay, I think we overall agree that this is horrible and should not be done this way, but it is the only way to do it right now.

On the other hand, as mentioned above, the function is really for a very specific case, and once a better solution comes around, we can just change it.